### ISSUE TITLE
[Enhancement] Something like C++ `requires`

### ISSUE THREAD

**Issue by disarray2077** [beefytech/Beef#1241]
I'm trying to hack my way through the lack of a ref constraint and came up with the below code, it looks correct but it doesn't work so I guess Beef doesn't support this kind of hackery, I have no idea if this is feasible but I think it could be quite useful if it did support it :)

Code with what I tried:
```cs
		public struct Yes;
		public struct No;

		public struct IsRef<T>
		{
			public typealias Result = comptype(_isref(typeof(T)));

			[Comptime]
			private static Type _isref(Type type)
			{
				if (let refType = type as RefType && refType.RefKind == .Ref)
					return typeof(Yes);
				return typeof(No);
			}
		}

		public struct DeRef<T>
		{
			public typealias Type = comptype(_derefType(typeof(T)));

			[Comptime]
			private static Type _derefType(Type type)
			{
				if (let refType = type as RefType && refType.RefKind == .Ref)
					return refType.UnderlyingType;
				return type;
			}
		}

		public static void Test1<T>()
			where T : var
			where IsRef<T>.Result : Yes
			where DeRef<T>.Type : struct
		{
			Console.WriteLine("Type is a struct (ref)");
		}

		public static void Test1<T>()
			where T : var, struct
		{
			Console.WriteLine("Type is a struct (not ref)");
		}

		public static void Test1<T>()
			where T : var
		{
			Console.WriteLine("Type is something else");
		}

		typealias refInt = rettype(function ref int());

		public static int Main(String[] args)
		{
			Compiler.Assert(IsRef(typeof(refInt)) == typeof(Yes)); // ok
			Compiler.Assert(DeRef(typeof(refInt)) == typeof(int)); // ok

			Test1<int>(); // ok
			Test1<refInt>(); // ERROR: 'no' must derive from 'yes'
			return 0;
		}
```

**Comment by bfiete** [comment]
`ref` isn't really part of the type system like in C++ - it's more like an attribute on variables and returns.

Either `rettype(function ref int())` should result in `int` or `Test1<refInt>()` should give an `invalid use of ref` error...

**Comment by bfiete** [comment]
Event<T> is currently broken, actually, then...
```
public rettype(T) Invoke(params T p) mut
{
	var result = default(rettype(T));
	for (var dlg in this)
	{
		result = dlg(params p);
	}
	return result;
}
```

Actually the only way that code would work at all would be if `rettype(T)` stripped `ref`. Otherwise: what would `default(ref int)` even mean? Since refs infer non-nullability, there's no possible default value.

**Comment by disarray2077** [comment]
> Event<T> is currently broken, actually, then...
Yeah, it fails to compile.
Looks like in C# they just throw a `NullReferenceException` if you try to call an unsubscribed event.

**Comment by disarray2077** [comment]
> ref isn't really part of the type system like in C++ - it's more like an attribute on variables and returns.

I see... that's problematic but that means that my "hack" should have worked if it wasn't about `ref`? I tried the following but it also didn't work, but this time for a different reason (the compiler chooses the wrong method)

```cs
		public struct Yes;
		public struct No;

		struct FakeRef<T>;

		public struct IsFakeRef<T>
		{
			public typealias Result = comptype(_isFakeRef(typeof(T)));

			[Comptime]
			private static Type _isFakeRef(Type type)
			{
				if (let refType = type as SpecializedGenericType && refType.UnspecializedType == typeof(FakeRef<>))
					return typeof(Yes);
				return typeof(No);
			}
		}

		public struct DeFakeRef<T>
		{
			public typealias Type = comptype(_deFakeRef(typeof(T)));

			[Comptime]
			private static Type _deFakeRef(Type type)
			{
				if (let refType = type as SpecializedGenericType && refType.UnspecializedType == typeof(FakeRef<>))
					return refType.GetGenericArg(0);
				return type;
			}
		}

		public static void Test3<T>()
			where IsFakeRef<T>.Result : Yes
			where DeFakeRef<T>.Type : var, struct
		{
			Console.WriteLine("Type is a struct (ref)");
		}

		public static void Test3<T>()
			where T : var, struct
		{
			Console.WriteLine("Type is a struct (not ref)");
		}

		public static void Test3<T>()
			where T : var
		{
			Console.WriteLine("Type is something else");
		}

		public static int Main(String[] args)
		{
			Compiler.Assert(typeof(IsFakeRef<FakeRef<int>>.Result) == typeof(Yes)); // ok
			Compiler.Assert(typeof(DeFakeRef<FakeRef<int>>.Type) == typeof(int)); // ok

			Test3<FakeRef<int>>(); // outputs: Type is a struct (not ref)
			return 0;
		}
```

**Comment by bfiete** [comment]
`ref` stripping at 8922cc0c0dcaed61dfdf9276b7ed9ca77c5c1a66
Ambiguous method call issue fixed at 12a3ba937a089caf4234aafa9c3651632b05052c

**Comment by disarray2077** [comment]
> `ref` stripping at [8922cc0](https://github.com/beefytech/Beef/commit/8922cc0c0dcaed61dfdf9276b7ed9ca77c5c1a66) Ambiguous method call issue fixed at [12a3ba9](https://github.com/beefytech/Beef/commit/12a3ba937a089caf4234aafa9c3651632b05052c)

Absolute poggers, this even allows what I suggested in issue https://github.com/beefytech/Beef/issues/1171 without code generation :D
(https://pastebin.com/tn3n9jMm)
